### DISCUSSION TITLE
Per-Shape Restitution and Friction

### DISCUSSION THREAD

**Discussion by LPGhatguy** [jrouwe/JoltPhysics#1311] [Q&A]
Hey! I'm part way through porting our existing game from [Rapier](https://rapier.rs) to Jolt.

With Rapier, you can set friction and restitution on a per-collider basis. A body can have multiple colliders, and each collider has one or more shapes. The result is that you can have a single rigid body that has different friction or restitution on just one part.

In Jolt, it looks like friction and restitution are set on the body instead of the shape. I think we could emulate what Rapier has by creating multiple rigid bodies and connecting them with constraints, but for rigidity reasons we're trying to merge assemblies into as few bodies as possible.

Is there a way to emulate the per-shape material properties we use in our Rapier implementation in Jolt?

**Comment by jrouwe** [comment]
There is an example that shows how to do this through a contact listener and a custom `PhysicsMaterial` (each shape of your compound could have a different material):

https://github.com/jrouwe/JoltPhysics/blob/master/Samples/Tests/General/FrictionPerTriangleTest.cpp

You can also use `ShapeSettings::mUserData` or `CompoundShapeSettings::SubShapeSettings::mUserData` or `CompoundShape::GetSubShapeIndexFromID` in combination with a contact listener. In this case you need to build a mapping between user data / sub shape index -> friction, restitution in your application code.
